Shallow Ice model

jabadge opened the pull request:
I added a new ice flow model to icepack, the shallow ice approximation. In icepack, the model is called shallow_ice. I have additionally created a test for the convergence of the shallow_ice diagnostic solver.

danshapero commented:
Thanks for submitting this! Looks good overall, I just have a few minor suggestions.

The tests seem to be failing because of something to do with gmsh. The testing environment that we use doesn't have gmsh installed (you can see the config for the testing environment [here](https://github.com/icepack/icepack-ci/blob/master/Dockerfile)). Let me update that and I'll get back to you once it's done. Basically this is all an oversight on my part.

danshapero commented:
@jabadge Can you rebase from master and force push your branch? I updated the testing environment so that it has gmsh now.

danshapero commented on icepack/models/shallow_ice.py:
```diff
@@ -0,0 +1,187 @@
+    return .5 * (inner(u,u))
+
+
+def part_2(u, h, s, A):
```
I think a more descriptive name like `gravity` would be better here -- in principle someone might want to modify this part of the model and that'll help cue them as to what it is.

danshapero commented on icepack/models/shallow_ice.py:
```diff
@@ -0,0 +1,187 @@
+# Split viscosity_gravity into two parts (with the first plus sign). Use the
+# first part to compute the scale because it is always positive.
+
+def part_1(u):
```
This could be renamed to `mass` or something more descriptive (see below).

danshapero commented on icepack/models/shallow_ice.py:
```diff
@@ -0,0 +1,187 @@
+# icepack source directory or at <http://www.gnu.org/licenses/>.
+
+import firedrake
+from firedrake import inner, grad, dx, ds
```
There's no boundary forcing, so you could leave off importing `ds`; likewise the water density isn't used here either.

danshapero commented on icepack/models/shallow_ice.py:
```diff
@@ -0,0 +1,187 @@
+            Ice thickness
+        s : firedrake.Function
+            Ice surface elevation
+        A : firedrake.
```
You could say `firedrake.Function or firedrake.Constant` for the type here.

danshapero commented on icepack/models/shallow_ice.py:
```diff
@@ -0,0 +1,187 @@
+
+        Returns
+        -------
+        E : firedrake.Function
```
Guessing this is a typo, the return type would be `firedrake.Form`.

danshapero commented on icepack/models/shallow_ice.py:
```diff
@@ -0,0 +1,187 @@
+        The positive part of the action functional is used as a dimensional
+        scale to determine when to terminate an optimization algorithm.
+        """
+        return self.part_1(u=u, **kwargs)
```
I think this needs a `* dx`? Also, the penalty term in the action is strictly positive, so we could include that in the scale functional as well.

jabadge replied:
Somehow I missed this comment in my initial changes. A new commit is coming. 
I moved the * dx to the action function as I think we had discussed. Do you want me to leave the * dx in the old location (as part of the mass, gravity, and penalty functions, as you have indicated here) or in the new location (where it currently is in the action function)? In the old location, it will be consistent with the rest of the models, but in the new location it will be consistent with changes you are about to make.

jabadge replied:
I added the penalty term to the scale, but I also realized that scale is not used in the shallow ice model. Should it be kept for future use or discarded?

danshapero replied:
The new location for all the measures is good. Good catch about the scale functional -- I think we should keep it for now. We can always remove it later. It might come in handy if, say, you want to add sliding.

danshapero replied:
I'm surprised the coverage checker hasn't chimed in yet, but if we do keep it, we should call the scale functional somewhere in the shallow ice test -- something as simple as

    assert firedrake.assemble(shallow_ice.scale(u=u, h=h, s=s, A=A) > 0)

would be plenty.

jabadge replied:
I added the penalty term to the scale function and included a test of the scale (ensuring that it is indeed positive) to the shallow_ice_test.py script.

danshapero commented on test/shallow_ice_test.py:
```diff
@@ -0,0 +1,106 @@
+# Copyright (C) 2017-2019 by Daniel Shapero <shapero@uw.edu>
```
Copyright should have your name, after all you wrote it :)

danshapero commented on icepack/models/shallow_ice.py:
```diff
@@ -0,0 +1,187 @@
+        """
+        return self.part_1(u=u, **kwargs)
+
+    def diagnostic_solve(self, u, h, s, A, **kwargs):
```
For consistency with the other models, can you swap `u` and `u0` here?

danshapero commented on test/shallow_ice_test.py:
```diff
@@ -0,0 +1,106 @@
+import firedrake
+from firedrake import inner, grad, dx, interpolate, as_vector, max_value
+import icepack, icepack.models
+from icepack.constants import (ice_density as ρ_I, water_density as ρ_W,
```
I think the water density and the Weertman sliding law imports can be removed for now.

danshapero commented on test/shallow_ice_test.py:
```diff
@@ -0,0 +1,106 @@
+    h_divide = (2 * R * (alpha/A0)**(1/n) * (n-1)/n)**(n/(2*n+2))
+    h_part2 = (n+1)*(r/R) - n*(r/R)**((n+1)/n) + n*(max_value(1-(r/R),0))**((n+1)/n) - 1
+    h_UFL = (h_divide/((n-1)**(n/(2*n+2)))) * (max_value(h_part2,0))**(n/(2*n+2))
+    return h_UFL
```
I've been using names like `h_expr` to denote something like what you have here as `h_UFL` -- something that's purely symbolic that will soon be interpolated into a function space. Can you change this for consistency with the other tests?

danshapero commented on test/shallow_ice_test.py:
```diff
@@ -0,0 +1,106 @@
+
+
+# Pick a geometry in order to have an exact solution. We'll use the
+# Bueler profile.
```
If people are reading the comments they might not know offhand what the Bueler profile is, so a reference to the right chapter and section of Greve and Blatter could be helpful.

danshapero commented:
Ok, this all looks great! And it looks like changing the domain radius increases the convergence rate, which is good to see.

The last thing is to rebase it down to one commit. This is to keep a clean and easily digestible commit history if we ever need to look back through it.

jabadge commented:
What should the one commit say? Do we want to keep a log of all the commits (as I did below) or should I summarize them? I tried rebasing to one commit on a test branch (rebase_squash_test). This is what I now see in the log:

commit 76f81323e393824395ed277d738c981b50fbda7f (HEAD -> rebase_squash_test)
Author: Jessica Badgeley <badgeley@uw.edu>
Date:   Mon Feb 3 09:51:53 2020 -0800
    
    Added the shallow ice approximation to model options.
    
    Added test for shallow ice model.
    
    Updated Shallow Ice model test to a circular mesh.
    
    Updated Shallow Ice model and test to match the formatting of the other models and tests.
    
    Made edits per Daniel Shapero's request before the pull request is accepted.
    
    Added option to ignore interpolation effects at the edges when testing shallow ice model.
    
    Added test to ensure the scale term in the shallow ice model is positive.

commit 01154fe0d94d84a635dd87a3aedb51a016b82d29 (origin/master, origin/HEAD, jabadge/master, master)
Author: Daniel Shapero <shapero.daniel@gmail.com>
Date:   Tue Apr 14 10:29:32 2020 -0700

    Updating testing environment
...

danshapero commented:
I usually keep the messages pretty short. The important part is that you added a shallow ice model and a test, which will be pretty self-explanatory to anyone who knows about ice flow. If someone needed to know more they can look at the actual diffs.

Sometimes I add more explanation if it isn't clear from the code itself. For example 62088cf was a bug fix related to things breaking in a dependency and in the commit message I explained a bit more about what went wrong. Without that extra context I'd have no idea why I did that if I looked back at it 6 months from now.
